pypeit/core/datacube.py

```diff
@@ -268,87 +315,101 @@ def extract_standard_spec(stdcube, subpixel=20):
+    # The 3D -> 2D conversion is done so that there is a spectral and spatial dimension,
+    # and the brightest white light pixel is transformed to be at the centre column of the 2D
+    # array. Then, the second brightest white light pixel is transformed to be next to the centre
+    # column of the 2D array, and so on. This is done so that the optimal extraction algorithm
+    # can be applied.
+    optkern = wl_img
+    if optfwhm is not None:
+        msgs.info("Generating a 2D Gaussian kernel for the optimal extraction, with FWHM = {:.2f} pixels".format(optfwhm))
+        x = np.linspace(0, wl_img.shape[0] - 1, wl_img.shape[0])
+        y = np.linspace(0, wl_img.shape[1] - 1, wl_img.shape[1])
+        xx, yy = np.meshgrid(x, y, indexing='ij')
+        # Generate a Gaussian kernel
+        intflux = 1
+        xo, yo = popt[1], popt[2]
+        fwhm2sigma = 1.0 / (2 * np.sqrt(2 * np.log(2)))
+        sigma_x, sigma_y = optfwhm*fwhm2sigma, optfwhm*fwhm2sigma
+        theta, offset, = 0.0, 0.0
+        optkern = gaussian2D((xx, yy), intflux, xo, yo, sigma_x, sigma_y, theta, offset).reshape(wl_img.shape)
+        # Normalise the kernel
+        optkern /= np.sum(optkern)
+
+    optkern_masked = optkern * mask[:,:,0]
+    # Normalise the white light image
+    optkern_masked /= np.sum(optkern_masked)
+    asrt = np.argsort(optkern_masked, axis=None)
+    # Need to ensure that the number of pixels in the object profile is even
+    if asrt.size % 2 != 0:
+        # Remove the pixel with the lowest kernel weight.
+        # It should be a zero value (given the mask), so it doesn't matter if we remove it
+        asrt = asrt[1:]
+    # Now sort the indices of the pixels in the object profile
+    tmp = asrt.reshape((asrt.size//2, 2))
+    objprof_idx = np.append(tmp[:,0], tmp[::-1,1])
+    objprof = optkern_masked[np.unravel_index(objprof_idx, optkern.shape)]
+
+    # Now slice the datacube and inverse variance cube into a 2D array
+    spat, spec = np.meshgrid(objprof_idx, np.arange(numwave), indexing='ij')
+    spatspl = np.apply_along_axis(np.unravel_index, 1, spat, optkern.shape)
+    # Now slice the datacube and corresponding cubes/vectors into a series of 2D arrays
+    numspat = objprof_idx.size
+    flxslice = (spatspl[:,0,:], spatspl[:,1,:], spec)
+    flxcube2d = _flxcube[flxslice].T
+    ivarcube2d = _ivarcube[flxslice].T
+    gpmcube2d = np.logical_not(bpmcube[flxslice].T)
+    waveimg = wave.reshape((numwave,1)).repeat(numspat, axis=1)
+    skyimg = np.zeros((numwave, numspat))  # Note, the residual sky has already been subtracted off _flxcube
+    oprof = objprof.reshape((1,numspat)).repeat(numwave, axis=0)
+    thismask = np.ones_like(flxcube2d, dtype=bool)
+
+    # Now do the optimal extraction
+    extract.extract_optimal(flxcube2d, ivarcube2d, gpmcube2d, waveimg, skyimg, thismask, oprof,
+                            sobj, min_frac_use=0.05, fwhmimg=None, base_var=None, count_scale=None, noise_floor=None)
```

Comment:
It's necessary in the sense that the `extract.extract_optimal()` currently works on 2D frames only. However, it need not be, but this would require a more significant refactor.